### ISSUE TITLE
Issues with Classic MIDI works Lighted Rocker Tabs

### ISSUE THREAD

**Issue by jefferycrowley** [GrandOrgue/grandorgue#1588]
Hello all!
I am having difficulty setting up Classic MIDI works Lighted Rocker Tabs with Grand Orgue.  This was working perfectly about a year ago and then I updated my Grand Orgue version and the rocker tabs do not work.   I had the same issue when upgrading from Hauptwerk V to VI.  Something has changed with the MIDI discovery.   It took me almost 4 months to finally get Hauptwerk working not only with my rocker tabs but also the LCD screens (3)O, Bar graphs (3) and Numeric readout for expression and crescendo visuals.   Previously in HW the rocker tabs took only a couple of minutes to setup correctly.   in HW VI the MIDI discovery allows the tab to be recognized but only as a BX controller.   I had to try every conceivable permutation (I have a long history with IT support and applications, and am retired so that I have all the time in the world to do this)  I found that in HW I had to "discover" the Lighted Rocker Tab (which I use for all main couplers, unison off and inter-manual coupling )  This also worked for Sweelinq application as well.  Then, I had to change what HW said was the device and modality used (in this case it was discovered as a BX Controller)  I tried all choices against other choices)  I finally got it to work correctly if I changed BX Controller to Common Controller Changes.  Using the  Receive tab  and had to change the controller type as 81-ON and 80-OFF in the Send Tab (this lights the tab up to say that it is "ON".   I have been able to configure correctly all of  the Indicator lights/dials eventually in Grand Orgue but it took me a couple of months of daily trial and error to get it to work.  I cannot get the LCD displays to work in Grand Orgue at all.   (Problem #2)  I use this for the Organ Name, the Temperament, The Combination set used, The last action used with the Combination buttons, the bank and position of the Sequencer).   My main focus will be with the Lighted Rocker Tabs.   When I attempt to MIDI discover the switches.  Gran Orgue finds the correct MIDI Channel in the receive tab, GO marks the object as a BX Controller and has the correct object -"MIDIworks Lighted Rocker Tab 003,004,0074,008 and 009 on the "Send" Tab.   I change the Send to 008 if the receive tab says 007 - which is how it was working a year ago.   I have even attempted to change the "Send" tab to the same LRT bank as the Receive Tab has (which is the thing that changed in Hauptwerk)  and also tried to change the BX Controller to Common Controller signals. and then on the "Send" tab I change it to be Control Change OFF instead of BX Controller and I put 81 in the "controller number" box and change the Receive tab to "Controller number" 80 - those numbers are for generic on/off switch or momentary controllers.   So it is setup the same way that Hauptwerk is set.   Also on the touchscreens the wrong couplers are drawn and/or many or all of the couplers turn on with the same tab.   Another tab shuts all or some of them off even though they have not been programmed yet and I checked the settings in all of the other couplers and they all were set to 'none' etc.    The other issue that I had about 6 months ago before I updated Grand Orgue was I could use the couplers, but the light wasn't triggered for the "ON" state and I had to press the same lower tab part to turn it off.  So "ON/OFF" had no light and have to use the same tab activity for both states.   I do not know how to setup the tabs.   Any help would be welcome!

**Comment by rousseldenis** [comment]
@jefferycrowley Could you just modify your message in order to add some spaces, indentations and carriage returns as it's just unreadable ?

Could you at the end, summarize topic by topic the problems you have ? Thanks

**Comment by oleg68** [comment]
@jefferycrowley 

1. How many Lighted Rocker Tabs do you have connected?
2. Could you post here the screenshot of the `File->Settings->MIDI Devices` (if the list of devices doesn't fit to the window, scroll them and make several screenshots)?
3. Please enable `Audio/MIDI->Log MIDI Events`, switch two tabs ON and OFF of each LRT and post here the event log (add comments, what tab are you pressed on each midi event).
4. Make "Detect complex MIDI setup" for some stops and press OK. Post here several screenshots of the Receive tabs.
5. Try to switch the tabs on LRT. Does GrandOrgue enables/disables the stops on the GUI panel?

**Comment by jefferycrowley** [comment]
[Grand Orgue Coupler - Rocker Tab Issues Submission  07122023.pdf](https://github.com/GrandOrgue/grandorgue/files/12034266/Grand.Orgue.Coupler.-.Rocker.Tab.Issues.Submission.07122023.pdf)

**Comment by oleg68** [comment]
@jefferycrowley You haven't answered my question and haven't provided the requested information.

**Comment by jefferycrowley** [comment]
> @jefferycrowley You haven't answered my questions and haven't provided the requested information.

Hello Oleg ~ I have answered your reply to me with a full PDF with screenshots and descriptions in the post just before this response.  It is the hyperlink entitled "Crand Orgue Coupler - Rocker Tab Issues Submissions07122023.pdf" Three days ago it is for issue #1588.

**Comment by oleg68** [comment]
@jefferycrowley

> Hello Oleg ~ I have answered your reply to me with a full PDF with screenshots and descriptions in the post 

You provided a very large PDF with lots of information that I hadn't requested, but without answering my questions. It is very hard to find any useful information there even it presented there.

Please post short numbered answers to the following points:

1. How many Lighted Rocker Tabs do you have connected?
2. Post here the screenshot of the File->Settings->MIDI Devices (if the list of devices doesn't fit to the window, scroll them and make several screenshots)?
3. Please enable Audio/MIDI->Log MIDI Events, switch two tabs ON and OFF of each LRT and post here the event log from the log window with your comments, what did you do before the event occured.
4. Make "Listen for event " for some stops and couplers and press OK. Post here several screenshots of the Receive tabs *without modifying anything manually*. Don't post here the Send tabs.
5. Repeat 4 with "Detect complex MIDI setup".

**Comment by jefferycrowley** [comment]
Oleg,

1. & 3.  I have 1 bank of 16 Lighted Rockers
Here is the requested Log MIDI Events for the rockers:
[MIDI Log for Rockers.txt](https://github.com/GrandOrgue/grandorgue/files/12065696/MIDI.Log.for.Rockers.txt)

2.  This is the File>MIDI Devices Page 1-8 and pages 19 & 20:
[GO Screenshots of MIDI Devices.pdf](https://github.com/GrandOrgue/grandorgue/files/12065803/GO.Screenshots.of.MIDI.Devices.pdf)

4. Listen to Event for only rocker Tabs as I do not have anything but Touch Screens for the Stops:
![I-P Listen to Event](https://github.com/GrandOrgue/grandorgue/assets/91970276/c2618f0e-c326-40fa-932b-6f371e799159)
![III-P Listen to Event](https://github.com/GrandOrgue/grandorgue/assets/91970276/4d06b541-f9a4-4cec-8510-5760d654600b)
![III-I Listen to Event](https://github.com/GrandOrgue/grandorgue/assets/91970276/b9a05c74-8d15-4159-a2de-08f536d4b63a)

And 5>  Detect Complex MIDI Event:
![I-P Detect complex MIDI Setup](https://github.com/GrandOrgue/grandorgue/assets/91970276/7fd536cc-4c72-49c6-805a-ed997598a8eb)
![III-P Detect complex MIDI setup](https://github.com/GrandOrgue/grandorgue/assets/91970276/8f5598bd-6799-4e8f-81ad-89b42cf47c2a)

**Comment by oleg68** [comment]
Thank you, @jefferycrowley for the information.

GrandOrgue intends that the rockers would send the same control number but different values for ON and OF of a key and different  control numbers of different rocker keys.

Instead, your Lighted Rockers send different control numbers (81 and 80) for ON and OFF events, but diferrent values for different keys. It is not supported nativelly by GrandOrgue now  (#1392 will improve it).

But you are still able to setup GrandOrgue for working with your Lighted Rocckers Tab with more tweaks: to assign two midi events to each switch.

1. Right Click on the GUI element you want to assign to a rocker key.
2. Go to Receive tab.
3. Be sure that there are no MIDI events already assigned to this element. If there are some, press the `Delete` button until you can not select more events in the `Event No` field.
4. Press 'Listen for Event' and switch your rocker key ON. A BxController event witt a controller no 81 should be discovered. 
    - *Don't change it to Bx Controller ON toggle*
    - The `Low Limit` and `Upper Limit` fields should be the same and should contain the rocker key number (from 0). Set them manually if they are wrong.
5. Press *`New`*, then press 'Listen for Event' and switch the rocker key OFF. A BxController event witt a controller no 80 should be discovered.
    - Correct the `Low Limit` and `Upper Limit` fields if it is necessary
6. Now you should have *two* different events set up in the `Receive` tab. You can switch between them with the `Event-no` field.
7. Press OK.
8. Check that the rocker key switches the GUI element correctly
9. Repeat the 1-8 steps for each key. You may to enter the values manually instead of using `Listen for Event`.

**Comment by jefferycrowley** [comment]
Hello Oleg:  Here are the results of setting my rocker tabs up for use with Grand Orgue.
[GO Rocker Tab - Results Submission.pdf](https://github.com/GrandOrgue/grandorgue/files/12073119/GO.Rocker.Tab.-.Results.Submission.pdf)

**Comment by oleg68** [comment]
@jefferycrowley Yes, you are right. The ``Ctrl Change Fixed...`` are right events for your case.
